df_engine_extensions/src/dist_sql_query/physical_plan.rs

```diff
@@ -169,19 +182,148 @@ impl<R: RemotePhysicalPlanExecutor> ExecutionPlan for ResolvedPartitionedScan<R>
+    type Item = DfResult<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        loop {
+            let stream_state = &mut this.stream_state;
+            match stream_state {
+                StreamState::Initializing => {
+                    let poll_res = this.stream_future.poll_unpin(cx);
+                    match poll_res {
+                        Poll::Ready(Ok(stream)) => {
+                            *stream_state = StreamState::Polling(stream);
+                        }
+                        Poll::Ready(Err(e)) => {
+                            *stream_state = StreamState::Failed(e.to_string());
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                        Poll::Pending => return Poll::Pending,
+                    }
+                }
+
+                StreamState::Polling(stream) => {
+                    let poll_res = stream.poll_next_unpin(cx);
+                    if let Poll::Ready(Some(Err(e))) = &poll_res {
+                        *stream_state = StreamState::Failed(e.to_string());
+                    }
+
+                    return poll_res;
+                }
+
+                StreamState::Failed(err_msg) => {
+                    return Poll::Ready(Some(Err(DataFusionError::Internal(format!(
+                        "failed to poll record stream, err:{err_msg}"
+                    )))));
+                }
+            }
+        }
+    }
+}
+
+/// Stream state
+/// Before polling record batch from it, we must initializing the record batch
+/// stream first. The process of state changing is like:
+///
+/// ```plaintext
+///         ┌────┐                                        
+///         │INIT│                                        
+///         └──┬─┘                                        
+///   _________▽_________                                 
+///  ╱                   ╲         ┌───────┐             
+/// ╱ Success to init the ╲________│POLLING│             
+/// ╲ record batch stream ╱yes     └────┬──┘             
+///  ╲___________________╱      ________▽_________        
+///            │no             ╱                  ╲       
+///        ┌───▽──┐           ╱ Success to poll    ╲___   
+///        │FAILED│           ╲ all record batches ╱yes│  
+///        └──────┘            ╲__________________╱    │  
+///                                     │no            │  
+///                                 ┌───▽──┐           │  
+///                                 │FAILED│           │  
+///                                 └──────┘           │  
+///                                                  ┌─▽─┐
+///                                                  │END│
+///                                                  └───┘
+/// ```
+pub enum StreamState {
+    Initializing,
+    Polling(DfSendableRecordBatchStream),
+    Failed(String),
```

Comment:
I guess there is no need to record the state of the failure of the inner stream, just let the inner stream to handle it failure:
```suggestion
    Initializing,
    InitializingFailed,
    Polling(DfSendableRecordBatchStream),
```